src/store_client.cc

```diff
@@ -702,13 +694,12 @@ storeUnregister(store_client * sc, StoreEntry * e, void *data)
         return 0;
     }
 
-    if (mem->clientCount() == 0) {
+    if (mem->clients.size() == 0) {
         debugs(90, 3, "storeUnregister: Consistency failure - store client being unregistered is not in the mem object's list for '" << e->getMD5Text() << "'");
         return 0;
     }
 
-    dlinkDelete(&sc->node, &mem->clients);
-    -- mem->nclients;
+    mem->clients.remove(sc);
```

Comment:
store_client's register and remove list entries asynchronously. So any attempt to hold an index will become invalid randomly.

We could rely on the CbcPointer magic and leave it in the registry. The "delete sc" shortly after will invalidate the registered Pointer and next time StoreEntry::invokeHandlers happens the list will be pruned. But that means no store_client related memory will get free'd until some time much later than currently happens.

Reply:
> any attempt to hold an index will become invalid randomly

The above statement contradicts C++ [guarantees](https://en.cppreference.com/w/cpp/container/list): "Adding, removing and moving the elements within the list or across several lists does not invalidate the iterators or references. An iterator is invalidated only when the corresponding element is deleted."

> We could rely on the CbcPointer magic and leave it in the registry.

We could, but an ever-growing list of nil pointers would create other problems/attack vectors. I see no good reason to create those problems so that we can solve them. Continuing to store the client index is a much simpler and safer solution AFAICT.